Python SWIG improvement

foleyj2 opened the pull request:
This started with me discovering issues in the debian package distribution of comedilib.
While trying to figure out what was wrong, I discovered that the way that backwards compatibility was implemented (in place globals() modification in an iterator)  broke Python3 compatibility, so I used copy.copy() to fix it.
I also reworked the backwards compatibility switches to make it easier to turn on and off using a #define SWIGPYTHONONLYSHORT switch.  (Incidentally, the Debian packages have it turned off at the moment.)

In addition, I have made test_comedi.py object-based, python2/3 compatible, mostly PEP8 compliant, and added two digital IO tests.  I haven't figured out how to make the command example happy on my Velleman P8061 board, but this follows with the caveat already mentioned in the previous version.  (Which did not compile on python2 or 3 on my computer.)

foleyj2 committed:
Ignore any python compiled files in SWIG directory: *.pyc

foleyj2 committed:
Fixed Python3 incompatibility created by trying to modify module globals() in iterator.  Also simplified breaking backwards compatibility switch as #define SWIGPYTHONONLYSHORT

foleyj2 committed:
Ignore __pycache__ created during python SWIG testing.

foleyj2 committed:
Renamed the python SWIG test program to reflect the focus on running a command.

foleyj2 committed:
Testing of the command routine is still broken, but digital IO works on my Velleman P8061-2.

ian-abbott reviewed (changes requested):
Except for trivial commit messages, please use a title line separated from the body of the commit message by a blank line. It looks better in the log that way.

foleyj2 commented:
Ah, I'll have to change my commit message formatting.  I don't think I did that on the new set of commits either.  I'll go fix them.  When will these changes make it into master?

foleyj2 commented:
Argh.  I didn't read your message about the commit message formatting before I started creating my updated documentation.  I'll have to delete the changes and re-commit them.

pieleric commented on swig/comedi.i:
```diff
@@ -115,14 +118,19 @@ unsigned int NI_AO_SCAN_BEGIN_SRC_RTSI(unsigned int rtsi_channel);
+import copy
 import re
-for k,v in globals().items():
+myglobals = copy.copy(globals())
```
You could make it just a little simpler by writing:
`for k,v in globals().copy().items():`

(And in turn this would avoid importing copy.copy and creating a new variable)

pieleric commented on swig/comedi.i:
```diff
@@ -115,14 +118,19 @@ unsigned int NI_AO_SCAN_BEGIN_SRC_RTSI(unsigned int rtsi_channel);
+for k,v in myglobals.items():
   if re.match('^comedi_', k, flags=re.IGNORECASE):
     globals()[k[7:]] = v
     # uncommenting following line removes compatibility with old code using
```
You could remove this comment about removing compatibility, as now there is a way to do so using `#define SWIGPYTHONONLYSHORT`.

pieleric commented on swig/python/test_comedi_command.py:
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python
```
Why renaming the file to `test_comedi_command.py` in this commit, and renaming it back to 'test_comedi.py` in the next commit?

ian-abbott replied:
This seems to be more than a simple rename, as the new file is twice as long as the old one. A better commit message is needed.

pieleric commented on swig/python/test_comedi_command.py:
```diff
@@ -0,0 +1,161 @@
+
+        dev = self.dev
+        # test our comedi command a few times.
+        ret = c.comedi_command_test(dev, cmd)
```
If you change the file to use the short function names, you should do it for all of them.
`comedi_command_test()` -> `command_test()`
`comedi_cmd_struct()` -> cmd_struct()`
....

pieleric commented on swig/python/test_comedi_command.py:
```diff
@@ -0,0 +1,161 @@
+                        help='comedi device, default /dev/comedi0')
+
+    ARGS = PARSER.parse_args()
+    CET = COMEDILIBtest(ARGS.input)
```
This is bit of nitpicking, but usually in Python, only constants are all upper-case. I'd recommend to use lowercase for parser, args, and cet.

pieleric reviewed (commented):
As the one who introduce the code to get rid of the long command names, I'm sorry I hadn't noticed that it broke Python 3 support. Indeed, in Python 3, `dict.items()` directly returns a view of the dict (instead of creating a new list), which means that modifying the dict while iterating over it causes a RuntimeError.

foleyj2 commented:
Hi Eric.

All of your comments seem quite relevant and I'll need to make such adjustments if I can figure out how.  I'm assuming I just commit and make a new pull request?  (I'm rather new to this git thing.)


I renamed test_comedi.py to test_comedi_command.py and back when I realized that it was easier to just make modular comprehensive test file rather than split it into multiple test files.

As for the uppercase variable names, that's due to my PEP8 checker complaining that they should be uppercase because it thinks they are constants at the top level:  C0103(invalid name).  I just re-read the naming convention in PEP8 and I guess it is getting confused.  They should be lowercase.

--

Dr. Joseph T. Foley, Assistant Professor, School of Science and Engineering, Reykjavik University +354-599-6569


________________________________
From: Éric Piel <notifications@github.com>
Sent: Monday, March 13, 2017 14:47
To: Linux-Comedi/comedilib
Cc: Joseph Timothy Foley; Author
Subject: Re: [Linux-Comedi/comedilib] Python SWIG improvement (#1)


@pieleric commented on this pull request.

As the one who introduce the code to get rid of the long command names, I'm sorry I hadn't noticed that it broke Python 3 support. Indeed, in Python 3, dict.items() directly returns a view of the dict (instead of creating a new list), which means that modifying the dict while iterating over it causes a RuntimeError.

________________________________

In swig/comedi.i<https://github.com/Linux-Comedi/comedilib/pull/1#discussion_r105675182>:

>  import re
-for k,v in globals().items():
+myglobals = copy.copy(globals())


You could make it just a little simpler by writing:
for k,v in globals().copy().items():

(And in turn this would avoid importing copy.copy and creating a new variable)

________________________________

In swig/comedi.i<https://github.com/Linux-Comedi/comedilib/pull/1#discussion_r105675693>:

>  import re
-for k,v in globals().items():
+myglobals = copy.copy(globals())
+# Copy dictionary before
+for k,v in myglobals.items():
   if re.match('^comedi_', k, flags=re.IGNORECASE):
     globals()[k[7:]] = v
     # uncommenting following line removes compatibility with old code using


You could remove this comment about removing compatibility, as now there is a way to do so using #define SWIGPYTHONONLYSHORT.

________________________________

In swig/python/test_comedi_command.py<https://github.com/Linux-Comedi/comedilib/pull/1#discussion_r105675892>:

> @@ -0,0 +1,161 @@
+#!/usr/bin/env python


Why renaming the file to test_comedi_command.py in this commit, and renaming it back to 'test_comedi.py` in the next commit?

________________________________

In swig/python/test_comedi_command.py<https://github.com/Linux-Comedi/comedilib/pull/1#discussion_r105676479>:

> +        cmd.start_src = c.TRIG_NOW
+        cmd.sart_arg = 0
+        cmd.scan_begin_src = c.TRIG_TIMER
+        cmd.scan_begin_arg = int(1.0E9 / 100000)
+        cmd.convert_src = c.TRIG_TIMER
+        cmd.convert_arg = 1
+        cmd.scan_end_src = c.TRIG_COUNT
+        cmd.scan_end_arg = nchans
+        cmd.stop_src = c.TRIG_COUNT
+        cmd.stop_arg = nscans
+        cmd.chanlist = mylist
+        cmd.chanlist_len = nchans
+
+        dev = self.dev
+        # test our comedi command a few times.
+        ret = c.comedi_command_test(dev, cmd)


If you change the file to use the short function names, you should do it for all of them.
comedi_command_test() -> command_test()
comedi_cmd_struct() -> cmd_struct()`
....

________________________________

In swig/python/test_comedi_command.py<https://github.com/Linux-Comedi/comedilib/pull/1#discussion_r105676879>:

> +        return ret
+
+if __name__ == "__main__":
+    BASEDIR = os.path.abspath(os.path.dirname(sys.argv[0]))
+
+    PARSER = argparse.ArgumentParser(
+        description='comedilib python SWIG command test',
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    # puts the defaults in the help automatically!
+    PARSER.add_argument('-d', '--debug', action='store_true',
+                        help='Set debug flags')
+    PARSER.add_argument('-i', '--input', default='/dev/comedi0',
+                        help='comedi device, default /dev/comedi0')
+
+    ARGS = PARSER.parse_args()
+    CET = COMEDILIBtest(ARGS.input)


This is bit of nitpicking, but usually in Python, only constants are all upper-case. I'd recommend to use lowercase for parser, args, and cet.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Linux-Comedi/comedilib/pull/1#pullrequestreview-26565257>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyfR1lOVRj00O_aqdBqsgUhApV_N-ddks5rlVcNgaJpZM4Matnl>.

pieleric commented:
Hi Joseph,
Typically, for updating a pull-request, you just update the branch from which you did the pull request (in this case, it's your master branch). A fancy way is to "rewrite history" of your repo, for instance by rebasing interactively and squashing commits, and then you can "git push --force". Github automatically detects that your branch has changed, and indicates it in the PR. Alternatively, you can just push additional commits on your branch, and they will appear in this PR automatically too.

Of course, deleting this PR, and making a new one, also works, but it looses the reviews, so it's not the recommended way.

foleyj2 commented:
Ah, unfortunately I didn't make a branch before I started making commits.  This is on the master of my fork.

Based upon what you said, I should make changes on master and they should show up in the PR?

--

Dr. Joseph T. Foley, Assistant Professor, School of Science and Engineering, Reykjavik University +354-599-6569


________________________________
From: Éric Piel <notifications@github.com>
Sent: Monday, March 13, 2017 20:20
To: Linux-Comedi/comedilib
Cc: Joseph Timothy Foley; Author
Subject: Re: [Linux-Comedi/comedilib] Python SWIG improvement (#1)


Hi Joseph,
Typically, for updating a pull-request, you just update the branch from which you did the pull request (in this case, it's your master branch). A fancy way is to "rewrite history" of your repo, for instance by rebasing interactively and squashing commits, and then you can "git push --force". Github automatically detects that your branch has changed, and indicates it in the PR. Alternatively, you can just push additional commits on your branch, and they will appear in this PR automatically too.

Of course, deleting this PR, and making a new one, also works, but it looses the reviews, so it's not the recommended way.

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Linux-Comedi/comedilib/pull/1#issuecomment-286231078>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyfR540OgRRU3EX_LpC-O50KkQuilkKks5rlaUVgaJpZM4Matnl>.

foleyj2 committed:
Added analog reads and writes.

Fixed the comedi SWIG calls to be using the shorter names consistently:  open() instead of comedi_open() as per Eric Piel suggestions.
Added analog read and write tests as methods read_analog() and write_analog().
Added print_function to better ensure Python3 compatibility.
Discovered that I was using uppercase letters in the main invocation because pylint was assuming they were constants.  I disabled that pylint invalid-name warning for the __main__ loop.

foleyj2 committed:
Fixed the "break backwards compatibility"

The lines
%rename("%(strip:[COMEDI_])s", regextarget=1) "COMEDI_.*";
%rename("%(strip:[comedi_])s", regextarget=1) "comedi_.*";
Caused errors on import due to _comedi function names not matching at import time.
Stripping names is now completely done in the globals() modification at the end, which works properly.
#ifdef SWIGPYTHONONLYSHORT sets a python variable "delete_comedi_prefix" to indicate if the old names should be stripped or not.
This makes the intent very clear in the generated comedi.py file.

foleyj2 commented:
I think I have now addressed all the points Eric brought up in the latest commits.

ian-abbott reviewed (commented):
> Testing of the command routine is still broken, but digital IO works on my Velleman P8061-2

The commit message should describe the changes made.

foleyj2 commented:
Oops.  How do I fix that?  That's an old commit message.

ian-abbott commented:
Some general advice.... When forking, it's best not to work on the 'master' branch directly (or rather, it's best not to work on the branches that you pull upstream changes into). Branches are cheap. Work on a new branch instead. When submitting a pull request and there are comments to be addressed, it is fairly straightforward to create a new version of the branch containing a cleaned-up version of the commits for a new pull request. (Including a version number in the branch name helps you to keep track of this.)

It's possible to turn your current 'master' branch into a feature branch retrospectively by checking out your current master into a new branch name and then resetting the master branch back to where it should be. Something like (on current master branch) `git checkout -b my-awesome-feature-v1` followed by `git branch -f master origin/master` (assuming the remote is called 'origin' and has a branch called 'master'). Because the master branch has only been reset locally, it will no longer match the master branch on github, but you can "force" push the master branch back to github to make it match: `git push --force origin master` (again, assuming the remote is called 'origin' and has a branch called 'master').

foleyj2 commented:
Thanks for the advice Ian.  I've never had to make a pull-request on github before.


I realized that I should have made a branch before making the pull-request.  Isn't it too late to fix this?  I can make a branch out of the commits, but the current pull-request is tied to the master branch on my fork.  I'd have to kill the current PR and resubmit, right?


Joe

--
Dr. Joseph T. Foley, Assistant Professor, School of Science and Engineering, Reykjavik University +354-599-6569


________________________________
From: ian-abbott <notifications@github.com>
Sent: Tuesday, March 14, 2017 12:25
To: Linux-Comedi/comedilib
Cc: Joseph Timothy Foley; Author
Subject: Re: [Linux-Comedi/comedilib] Python SWIG improvement (#1)


Some general advice.... When forking, it's best not to work on the 'master' branch directly (or rather, it's best not to work on the branches that you pull upstream changes into). Branches are cheap. Work on a new branch instead. When submitting a pull request and there are comments to be addressed, it is fairly straightforward to create a new version of the branch containing a cleaned-up version of the commits for a new pull request. (Including a version number in the branch name helps you to keep track of this.)

It's possible to turn your current 'master' branch into a feature branch retrospectively by checking out your current master into a new branch name and then resetting the master branch back to where it should be. Something like (on current master branch) git checkout -b my-awesome-feature-v1 followed by git branch -f master origin/master (assuming the remote is called 'origin' and has a branch called 'master'). Because the master branch has only been reset locally, it will no longer match the master branch on github, but you can "force" push the master branch back to github to make it match: git push --force origin master (again, assuming the remote is called 'origin' and has a branch called 'master').

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Linux-Comedi/comedilib/pull/1#issuecomment-286405732>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyfR02sdMtwIdxMM1hIz7AviX2KKCZ8ks5rlodNgaJpZM4Matnl>.

ian-abbott commented:
Yes, you can resubmit a PR on a different branch. It's worth including a version number in the name of the submitted branch so you can keep track of it easier.

Also, try and wrap the body of the commit messages to 72 columns where possible (the title line and code snippets included in the commit message might be longer), and include a short tag at the start of the patch title where appropriate, although there are no strict rules for these tags. Look at some of the earlier commit messages to see what I mean. These commit messages eventually end up in the plain text 'ChangeLog' file in the distributed tar files when a new version is released, and long paragraphs without line-breaks make it look ugly!

foleyj2 commented:
So, is there anything else to be done on this PR before it can be accepted?  I'll make a branch for the next PR I make (which will be about updating the documentation.)

ian-abbott commented:
> So, is there anything else to be done on this PR before it can be accepted? I'll make a branch for the next PR I make (which will be about updating the documentation.)

I can fix up the messy history, but I could do with descriptions of what has actually been done to test_comedi.py in commits 2666afdd and dc788354.

ian-abbott commented:
I've merged your changes to `swig/comedi.i` into master (with some edits to the commit messages), but I've dropped the `swig/python/test_comedi.py` script altogether as the original never worked, and your changes were a bit too device-specific.

It will probably be better to have some small, generic demo programs similar to the 'C' comedilib demo programs.

foleyj2 commented:
Thank you.  I meant to get around to doing the cleanup, but I didn’t.
I’ll try to come up with a generic demo program, unfortunately I only had the Velleman board for comparison.

From: ian-abbott [mailto:notifications@github.com]
Sent: Monday, April 10, 2017 13:44
To: Linux-Comedi/comedilib <comedilib@noreply.github.com>
Cc: Joseph Timothy Foley <foley@ru.is>; Author <author@noreply.github.com>
Subject: Re: [Linux-Comedi/comedilib] Python SWIG improvement (#1)


I've merged your changes to swig/comedi.i into master (with some edits to the commit messages), but I've dropped the swig/python/test_comedi.py script altogether as the original never worked, and your changes were a bit too device-specific.

It will probably be better to have some small, generic demo programs similar to the 'C' comedilib demo programs.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Linux-Comedi/comedilib/pull/1#issuecomment-292953315>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAyfR6dIk9Ps9nFn-RgomhgzBa8pImfzks5rujI4gaJpZM4Matnl>.
